src/upgrader_canister/src/state/polls.rs

```diff
@@ -76,7 +97,61 @@ impl<M: Memory> Polls<M> {
             poll.no_voters.push(voter_principal);
         }
 
-        self.polls.insert(poll_id, poll);
+        self.pending_polls.insert(poll_id, poll);
+        Ok(())
+    }
+
+    /// Finalizes the poll and moves it to the closed polls
+    pub fn finalize_polls(
+        &mut self,
+        timestamp_secs: u64,
+        permissions_service: &mut Permissions<M>,
+    ) -> Result<()> {
+        // loop through all the pending polls and find the closed ones
+        let mut polls_to_close = Vec::new();
+        for (id, poll) in self.pending_polls.iter() {
+            if timestamp_secs > poll.end_timestamp_secs {
+                polls_to_close.push((id, poll.clone()));
+            }
+        }
+
+        // close the polls
+        for (id, poll) in polls_to_close {
+            self.process_poll(&poll, permissions_service)?;
+            self.pending_polls.remove(&id);
+            self.closed_polls.insert(id, poll);
+        }
+
+        Ok(())
+    }
+
+    /// Process a pool before it is finalized
+    pub fn process_poll(
```

Comment:
this function is not too clear to me. Is it a function to apply the poll results? In this case, the name should remind that. Also, you could avoid nesting by checking at first, and returning Ok if no voters is bigger than yes voters.

Also, even if the code is very short, I would move the effect for each poll type to a separate private function.

Another thing not clear to me is: why doesn't this function check whether the poll is finished? It's just checking the number of votes. Maybe I'm missing a part of the flow.

Reply:
This is going to be refactored